pkg/source/adapter/adapter.go

```diff
@@ -208,11 +227,96 @@ func (a *Adapter) SetRateLimits(r rate.Limit, b int) {
 	a.rateLimiter = rate.NewLimiter(r, b)
 }
 
-func (a *Adapter) HandleServiceMessage(ctx context.Context, message ctrl.ServiceMessage) {
-	// In this first PR, there is only the RA sending messages to control plane,
-	// there is no message the control plane should send to the RA
-	a.logger.Info("Received unexpected control message")
+func (a *Adapter) HandleSetContract(ctx context.Context, message ctrl.ServiceMessage) {
+	newContract := kafkasourcecontrol.KafkaSourceContract{}
+	if err := newContract.UnmarshalBinary(message.Payload()); err != nil {
+		a.logger.Errorf("Catastrophic failure: received an unparsable contract. This sounds like a programming bug: %v", err)
+		message.Ack()
+		return
+	}
+
+	// We have the contract, we can ack back now
 	message.Ack()
+
+	// Let's lock and proceed with the changes
+	a.reconfigurationLock.Lock()
+	defer a.reconfigurationLock.Unlock()
+
+	if reflect.DeepEqual(&newContract, a.actualContract) {
+		// Nothing to do here
+		return
+	}
+
+	consumerContext, stopConsumer := context.WithCancel(ctx)
+	a.stopConsumer = stopConsumer
+
+	a.stopConsumerGroup()
+	a.actualContract = &newContract
+	a.consumerStoppedSignal = a.startConsumerGroup(consumerContext, consumer.WithSaramaConsumerLifecycleListener(a))
+}
+
+func (a *Adapter) stopConsumerGroup() {
+	if a.stopConsumer != nil {
+		a.stopConsumer()
+	}
+	if a.consumerStoppedSignal != nil {
+		<-a.consumerStoppedSignal.Done()
+	}
+}
+
+func (a *Adapter) startConsumerGroup(ctx context.Context, consumerGroupOptions ...consumer.SaramaConsumerHandlerOption) context.Context {
+	// I'm overwriting these to avoid creating a too big change here
+	// Note: Some parts of the config still comes from the env, hence this can't be trivially removed
+	// TODO(slinkydeveloper) make client.NewConfig flexible to accept KafkaSourceContract too
```

Comment:
as part of the cfg ?

Reply:
I still don't know how to solve this TODO, will get back on it when I have the error propagation in place